sampler_test.go

```diff
@@ -575,3 +578,106 @@ func TestRemotelyControlledSampler_updateSamplerFromAdaptiveSampler(t *testing.T
 	_, ok = remoteSampler.sampler.(*rateLimitingSampler)
 	require.True(t, ok)
 }
+
+func TestAdaptiveSampler_lockRaceCondition(t *testing.T) {
+	agent, remoteSampler, _ := initAgent(t)
+	defer agent.Close()
+	remoteSampler.Close() // stop timer-based updates, we want to call them manually
+
+	numOperations := 100
+	strategiesA := generatePerOperationStrategies(numOperations, "a")
+	strategiesB := generatePerOperationStrategies(numOperations, "b")
+	adaptiveSampler, err := NewAdaptiveSampler(strategiesA, 200)
+	require.NoError(t, err)
+
+	// Overwrite the sampler with an adaptive sampler
+	remoteSampler.sampler = adaptiveSampler
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	wg.Add(3)
+
+	exit1 := make(chan struct{})
+	exit2 := make(chan struct{})
+	exit3 := make(chan struct{})
+
+	go func() {
+		defer wg.Done()
+		isSampled(t, exit1, remoteSampler, numOperations, "a")
+	}()
+	go func() {
+		defer wg.Done()
+		isSampled(t, exit2, remoteSampler, numOperations, "b")
+	}()
+	go func() {
+		defer wg.Done()
+		updateSampler(
+			exit3,
+			remoteSampler,
+			agent,
+			&sampling.SamplingStrategyResponse{OperationSampling: strategiesA},
+			&sampling.SamplingStrategyResponse{OperationSampling: strategiesB},
+		)
+	}()
+	time.Sleep(10 * time.Second)
+
+	close(exit1)
+	close(exit2)
+	close(exit3)
+}
+
+func isSampled(t *testing.T, exit chan struct{}, remoteSampler *RemotelyControlledSampler, numOperations int, operationNamePrefix string) {
+	for {
+		select {
+		case <-exit:
+			return
+		default:
+			sampled, _ := remoteSampler.IsSampled(TraceID{}, generateRandomOperationName(numOperations, operationNamePrefix))
+			assert.True(t, sampled)
+			time.Sleep(10 * time.Microsecond)
+		}
+	}
+}
+
+func updateSampler(
```

Comment:
can you explain better what this is doing? It looks like it's constantly causing adaptive sampler to discard half of the endpoints.